### DISCUSSION TITLE
What is the proper way to assign stereochemistry for bonds with the new CIP system?

### DISCUSSION THREAD

**Discussion by shubbey** [rdkit/rdkit#5367] [Q&A]
Following up on this report: https://github.com/rdkit/rdkit/issues/5357

I am testing RDKit to potentially replace our in-house CIP-chirality algorithm which only handles CIP rules 1-3 (of 6).  However, I am having trouble getting it to work properly particularly when stereo on bonds is concerned.  I'm pretty sure there is some pre-processing call that I am missing here.  

For reference, I'm validating rules by running RDKit against the molecules here: https://cipvalidationsuite.github.io/ValidationSuite/

```
from rdkit import Chem
def compareCIPs(rdmol,name,truth):
    #  what other pre-processing goes here?
    Chem.rdCIPLabeler.AssignCIPLabels(rdmol)
    ainfos = []
    for i,rda in enumerate(rdmol.GetAtoms()):
        if rda.HasProp("_CIPCode"):
           cip = rda.GetProp("_CIPCode")
           if cip:
               ainfos.append((i+1,cip))

    for i,rda in enumerate(rdmol.GetBonds()):
        if rda.HasProp("_CIPCode"):
           cip = rda.GetProp("_CIPCode")
           if cip:
               atom1 = rda.GetBeginAtomIdx();
               atom2 = rda.GetEndAtomIdx();
               ainfos.append((atom1+1,cip))
               ainfos.append((atom2+1,cip))

    ainfos.sort();
    pred = ""
    first = False
    for ainfo in ainfos:
        if first:
            pred += " "
        first = True
        pred += str(ainfo[0])+ainfo[1]

    if(pred != truth):
        #print(name)
        #print(truth)
        #print(pred)
        return 1
    return 0

def checkSmiles(smilesFile):
    with open(smilesFile,'r') as f:
        lines = f.readlines()

    fail = 0
    for line in lines:
        words = line.split('\t')
        smiles = words[0]
        name = words[1]
        labels = words[2]
        rdmol = Chem.MolFromSmiles(smiles,sanitize=False) 
        Chem.SanitizeMol(rdmol) #  necessary to avoid valence errors
        fail += compareCIPs(rdmol,name,labels)
    return fail,len(lines)

def checkSDF(sdfFile):
    fail = 0
    suppl = Chem.SDMolSupplier(sdfFile,sanitize=True,removeHs=False,strictParsing=False)
    # <- generate stereo here??
    for rdmol in suppl:
        name = rdmol.GetProp("STRUCTURE_ID")
        labels = rdmol.GetProp("CIP_LABELS") if rdmol.HasProp("CIP_LABELS") else ""
        fail += compareCIPs(rdmol,name,labels)
    return fail,len(suppl)

fail,tot = checkSmiles('compounds.smi')
print('Smiles:',fail,'/',tot,'failed')
fail,tot = checkSDF('compounds_3d.sdf')
print('SDF:',fail,'/',tot,'failed')
```

Which gives:
```
python chiral.py
Smiles: 88 / 300 failed
SDF: 62 / 300 failed
````

Investigating these in more detail shows that the majority of the failures are due to missing E/Z labels.  But I haven't been able to figure out which calls to try to generate these.  From digging under the hood, it appears that Chem.FindPotentialStereo() is used for the newer CIP, but I have tried this as well as Chem.FindPotentialStereoBonds() and neither changes the results.  For the SDF import, where coords are present, I've tried Chem.AssignStereochemistryFrom3D() but that too did nothing.  

Any advice?  Thanks!

**Comment by greglandrum** [comment]
Hi @shubbey: double bonds which have stereo set should end up with a `_CIPCode` property set:
```
In [4]: m = Chem.MolFromSmiles('C/C=C/C')

In [5]: from rdkit.Chem import rdCIPLabeler

In [6]: rdCIPLabeler.AssignCIPLabels(m)

In [7]: m.GetBondWithIdx(1).GetPropsAsDict()
Out[7]: {'_CIPCode': 'E'}

```

**Comment by shubbey** [comment]
Thanks, Greg.

This may be better filed as a bug report then because I am getting several failures from the above validation set when loading from Smiles as you do above.  Example:

VS063: O[C@H]/1CCCCCCCCCCCCCCCCCCCCCCC[C@H]2CC1=C=C=C/CCCCCCCCCCCCCCCCC2
https://chemapps.stolaf.edu/jmol/jsmol/cip.htm?load=https://cipvalidationsuite.github.io/ValidationSuite/mol3d/VS063.sdf
(RDKit does not make the 'E' assignment)

From running the code above, with the change to sanitize within the Smiles loader, I get 55/300 failures.  These are divided as so:

R/S failures:  VS132 , VS246-248, VS279-300
r/s failures: VS207, VS214, VS226, VS231
E/Z failures: VS063, VS118, VS135, VS154, VS164, VS188, VS189, VS190, VS214
e/z failures: VS246-248, VS 299

As well as a bunch of M/P failures but I'm not concerned with those right now (it's unclear whether the CIP code is even outputting this marker but they are labeled in the compounds.smi file).  

Am I correct that this Centres-ported CIP code is supposed to cover all of the currently known CIP rules?

Best,
shubbey

**Comment by shubbey** [comment]
Just FYI, I now see in the CIPLabeler README file that Centres isn't completely ported, specifically some of the more esoteric stereochemical forms.  So that probably explains the issues here.
